content/docs/tutorials/serialmonitor.md

```diff
@@ -0,0 +1,294 @@
+            if c < 32 {
+                // Convert nonprintable control characters to
+                // ^A, ^B, etc.
+                machine.Serial.WriteByte('^')
+                machine.Serial.WriteByte(c + '@')
+            } else if c >= 127 {
+                // Anything equal or above ASCII 127, print ^?.
+                machine.Serial.WriteByte('^')
+                machine.Serial.WriteByte('?')
+            } else {
+                // Echo the printable character back to the
+                // host computer.
+                machine.Serial.WriteByte(c)
+            }
+        }
+
+        // This assumes that the input is coming from a keyboard
+        // so checking 120 per second is sufficient. But if the
+        // data comes from another processor, the port can
+        // theoretically receive as much as 11000 bytes/second
+        // (115200 baud). This delay can be removed and the
+        // Serial.Read() method can be used to retrieve
+        // multiple bytes from the receive buffer for each
+        // iteration.
+        time.Sleep(time.Millisecond * 8)
+    }
+}
+```
+
+You can flash this program to the microcontroller (in this example, a SAMD21 M0+
+clone that emulates an Arduino Zero), and fire up the monitor like this:
+
+```
+$ tinygo flash -target=arduino-zero
+$ tinygo monitor
+Connected to /dev/ttyACM0. Press Ctrl-C to exit.
+Reading from the serial port...
+abcdef^A^B^D^E^F^G^H^I^J^K^L^M^N^O^P^R^T^U^V^W^X^Y^^^[^]^_^?
+```
+
+Type a few characters in the `tinygo monitor`, for example "abcdef". You should
+see the characters echoed back by the microcontroller, as shown above. If you
+type a [nonprintable control
+characters](https://en.wikipedia.org/wiki/C0_and_C1_control_codes), these are
+echoed back as 2 characters: the caret character `^` and a letter representing
+the control character. For example, typing Control-P prints `^P`.
+
+A number of control characters are intercepted by the `tinygo monitor` itself
+instead of being sent to the microcontroller:
+
+* Control-C: terminates the `tinygo monitor`
+* Control-Z: suspends the `tinygo monitor` and drops back into shell
+* Control-\\: terminates the `tinygo monitor` with a stack trace
+* Control-S: flow-control, suspends output to the console
+* Control-Q: flow-control, resumes output to the console
+* Control-@: thrown away by `tinygo monitor`
+
+## Alternative Serial Monitors
+
+There are many alternative serial monitor programs that can be used instead of
+`tinygo monitor`. The setup is slightly more complicated because you need to
+know the serial port on the host computer that the microcontroller is mapped to.
+
+On Linux machines, the microcontrollers will be assigned a serial port that has
+a `USB` prefix or an `ACM` prefix like this:
+
+* `/dev/ttyUSB0`
+* `/dev/ttyACM0`
+
+On MacOS machines, the serial port will look like this:
+
+* `/dev/cu.usbserial-1420`
+* `/dev/cu.usbmodem6D8733AC53571`
+
+[TODO: No idea on Windows.]
```

Comment:
On Windows, strings such as `COM1` and `COM31` are used as port names.

Reply:
Added `COM1` and `COM31`. Thanks.

I didn't know about the default being 9600 on AVR processors. 
* Added `tinygo monitor -baudrate=9600` for AVR.
* Added paragraph explaining which processors use 115200 and which use 9600.

Is there a reason why 115200 doesn't work on AVR? When I program them with the Arduino IDE/CLI, I always use 115200 for the serial monitor, and I haven't run into any issues.

Fixed a couple of grammar typos.

I am reluctant to add the `-monitor` flag in this tutorial. Because it does not work consistently. The monitor starts too quickly after flashing, and it terminates with an error message like this:
```
$ tinygo flash -target=arduino-zero -monitor
[...]
Verify successful
Done in 0.483 seconds
Connected to /dev/ttyACM0. Press Ctrl-C to exit.
error: read error: Port has been closed
```
The `-monitor` flag seems to work only for microcontroller boards using a USB-to-serial adapter, which presumably keeps the serial port connection alive during the flashing and reboot process.

I do mention the `-monitor` flag in the subcommand reference that I added in PR #330.

Reply:
The AVR does not necessarily need to be 9600 bps.
However, since TinyGo defaults to 9600 bps, I believe the `-baudrate` flag needs to be explained.

https://github.com/tinygo-org/tinygo/blob/v0.27.0/src/machine/machine_atmega.go#L161-L165

Reply:
> I am reluctant to add the -monitor flag in this tutorial. Because it does not work consistently. The monitor starts too quickly after flashing, and it terminates with an error message like this:

I added tinygo monitor and I am aware that it does not work with some targets.
However, it works fine on my samd21/samd51/rp2040/nrf52840.

* samd21 : xiao / feather-m0
* samd51 : feather-m4 / wioterminal
* rp2040 : pico / xiao-rp2040
* nrf52840 : feather-nrf52840 / xiao-ble

And it works better with USBtoSerial.

I believe the case of failure as in your example is the case of a slow reboot of the microcontroller after flashing.
tinygo monitor accidentally connects to bootloader and disconnects at reboot after flash.
If you are using UF2 bootloader, changing to a newer version should solve many of your problems.

https://github.com/microsoft/uf2-samdx1

Reply:
Please take a look at the latest version. I think I incorporated most (if not all) of your comments. The behavior of `tinygo monitor` when there are multiple boards attached is not consistent. Sometimes it can figure out the port, sometimes it cannot, and sometimes `flash` will figure it out, but `-monitor` does not. I tried to explain this without overwhelming the beginner...

Reply:
Thank you very much. I think it is a very good and easy to understand document.